### PR TITLE
Travis CI - Add support for runtime system operational test

### PR DESCRIPTION
**Pull request by tomereli**
CI support for run-time operational testing using the docker runner which runs prplmesh controller and agent in one container, then uses docker exec command to call prplmesh_utils.sh status which fails if not in operational mode, thus failing CI.

### CODE REVIEW

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 1]
```diff
@@ -136,19 +140,46 @@ stop_function() {
     [ "$PRPLMESH_MODE" = "CA" -o "$PRPLMESH_MODE" = "A" ] && prplmesh_agent_stop
 }
 
+main_agent_operatioanl() {
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: typo: operatioanl -> operational

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 2]
```diff
@@ -136,19 +140,46 @@ stop_function() {
...
+}
+
 status_function() {
     echo "$0: status"
     ps -aux | grep beerocks | grep -v grep
     ps -aux | grep ieee1905_transport | grep -v grep
     ps -aux | grep local_bus | grep -v grep
+
+    # check for operational status
+    if [ "$PRPLMESH_USER" != "" ]; then
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: PRPLMESH_USER is not used, and I don't see it's purpose.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 3]
```diff
@@ -136,19 +140,46 @@ stop_function() {
...
+        LOGS_PATH="/tmp/$PRPLMESH_USER/beerocks/logs/"
+    else
+        LOGS_PATH="@BEEROCKS_TMP_PATH@/logs/"
+    fi
+
+    error=false
+    main_agent_operatioanl $LOGS_PATH && success "Main agent operational" || {
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: operational again

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 4]
```diff
@@ -136,19 +140,46 @@ stop_function() {
...
+    }
+    radio_agent_operational $LOGS_PATH wlan2 && success "wlan2 radio agent operational" || {
+        err "wlan2 radio agent not operational"; error=true
+    }
+    return $error
 }
 
 usage() {
-    echo "usage: $(basename $0) {start|stop|restart|status} [-hvsm]"
+    echo "usage: $(basename $0) {start|stop|restart|status} [-hvsmu]"
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: Not needed.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 5]
```diff
@@ -136,19 +140,46 @@ stop_function() {
...
 }
 
 main() {
-    OPTS=`getopt -o 'hvm:s' --long verbose,help,mode:skip-platform -n 'parse-options' -- "$@"`
+    OPTS=`getopt -o 'hvm:u:s' --long verbose,help,mode:skip-platform -n 'parse-options' -- "$@"`
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: Not needed.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 6]
```diff
@@ -159,6 +190,7 @@ main() {
             -v | --verbose)       VERBOSE=true; shift ;;
             -h | --help)          usage; exit 0; shift ;;
             -m | --mode)          PRPLMESH_MODE="$2"; shift; shift ;;
+            -u | --user)          PRPLMESH_USER="$2"; shift; shift ;;
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: Not needed.

File: .travis.yml [hunk 7]
```diff
@@ -10,6 +10,10 @@ script:
   - cd tools/docker
```
Comment (arnout) [inline .travis.yml:10]: type in commit message: continues -> continuous

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 8]
```diff
@@ -136,19 +140,46 @@ stop_function() {
...
+    main_agent_operatioanl $LOGS_PATH && success "Main agent operational" || {
+        err "Main agent not operational"; error=true
+    }
+    radio_agent_operational $LOGS_PATH wlan0 && success "wlan0 radio agent operational" || {
+        err "wlan0 radio agent not operational"; error=true
+    }
+    radio_agent_operational $LOGS_PATH wlan2 && success "wlan2 radio agent operational" || {
+        err "wlan2 radio agent not operational"; error=true
+    }
+    return $error
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in:184]: Should be either just
```
$error
```
(running `true` or `false`, and the last command of the function gives the exit code), or
``
$error || return 1
```

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 3 above)
Comment (arnout) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: `&&` and `||` order of evaluation is not portable. Better use an explicit `if` condition.

This could also be factored in a function
```
report() {
    msg="$1"; shift
    if "$@"; then
        success " OK  $msg"
    else
        err "FAIL $msg"
        error=true
    fi
}
...
    report "Main agent operational" \
        main_agent_operational $LOGS_PATH
```

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 9]
```diff
@@ -161,15 +161,21 @@ status_function() {
         LOGS_PATH="@BEEROCKS_TMP_PATH@/logs/"
     fi
 
-    error=false
+    error=0
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in:173]: This should already have been done in the first commit.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 10]
```diff
@@ -161,15 +161,21 @@ status_function() {
...
+    cat $LOGS_PATH/beerocks_agent.log
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: This is not nice.It would be better to make sure the logs are mapped to the host and export them as build artifacts for later inspection. But that's for some other day. :ok:

File: tools/docker/README.md [hunk 11]
```diff
@@ -22,7 +22,7 @@ The docker builder image can be used to build prplMesh via `maptools.py build` c
 Build prplMesh in container:
 
 ```bash
-sudo ./builder/build.sh <maptools.py build arguments>
+sudo ./build.sh <maptools.py build arguments>
```
Comment (arnout) [inline tools/docker/README.md:25]: Why this change?

File: tools/docker/README.md [hunk 12]
```diff
@@ -33,6 +33,10 @@ The runner docker allows running multiple containers, for example one with a con
 The 2 containers are connected using a docker network (bridge), so can
 communicate + sniffed by running `wireshark` / `tcpdump` on the bridge from the host to see 1905 packets.
 
+## Docker tester
+
+A simple wrapper which calls `sudo docker exec <container name> -it <path/to>/build/install/scripts/prplmesh_utils.sh status which prints the prplMesh operational status of the main and radio agents.
```
Comment (arnout) [inline tools/docker/README.md]: Missing closing backquote.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 13]
```diff
@@ -8,11 +8,11 @@ dbg() {
 }
 
 err() {
-	echo -e '\033[1;31m'"$@"'\033[0m'
+    echo '\033[1;31m'"$@"'\033[0m'
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: Should be printf, echo will not necessarily expand escapes.

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 14]
```diff
@@ -37,12 +37,12 @@ prplmesh_platform_db_init() {
     management_mode=${1-Multi-AP-Controller-and-Agent}
     operating_mode=${2-Gateway}
 
-    mkdir -p @PLATFORM_DB_PATH_TEMP@/
+    mkdir -p /tmp/
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in:40]: Rebase screw-up?

File: .travis.yml [hunk 15]
```diff
@@ -10,6 +10,10 @@ script:
...
   - sudo ./image-build.sh -b ubuntu:18.04
   - sudo ./build.sh map -f MSGLIB=zmq BUILD_TESTS=ON
+  # Start controller & agent node detached and check status
+  - sudo ./run.sh start-controller-agent -d -n controller
```
Comment (arnout) [inline .travis.yml]: Since it has controller and agent, it would be better to use `-n controller-agent`.

File: tools/docker/test.sh [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+run() {
+    echo "$*"
+    "$@" || exit $?
+}
+
+scriptdir="$(cd "${0%/*}"; pwd)"
+topdir="${scriptdir%/*/*/*}"
+
+main() {
+    run docker container exec -it ${1-agent} ${topdir}/build/install/scripts/prplmesh_utils.sh status
```
Comment (arnout) [inline tools/docker/test.sh]: It would be better to use the -n option for the name, to make it similar to the run.sh.

**Review by arnout — CHANGES_REQUESTED** [review]

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 3 above)
Comment (tomereli) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: But error is local variable so how can I use it after the report? I want to dump the log on failure.

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 9 above)
Comment (tomereli) [reply common/beerocks/scripts/prplmesh_utils.sh.in:173]: changed back to false / true following your previous comment

File: tools/docker/README.md (hunk 11 above)
Comment (tomereli) [reply tools/docker/README.md:25]: Ostap's - I moved the build.sh, run.sh to the docker directory in one of the previous commits not part of this PR and forgot to change the README. the builder/runner contain now the images only.

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 14 above)
Comment (tomereli) [reply common/beerocks/scripts/prplmesh_utils.sh.in:40]: no, fixing another issue. I didn't know you will be reviewing, I should have opened a draft PR. But its a good thing you review this now

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 3 above)
Comment (arnout) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: shell doesn't have local variables.

bash has local variables, but only if explicitly declared with `local`.

You also get a kind of local variable by doing things in a subshell with `(...)`.

File: .travis.yml [hunk 17]
```diff
@@ -8,8 +8,14 @@ services:
 
 script:
   - cd tools/docker
-  - sudo ./image-build.sh -b ubuntu:18.04
+  - sudo ./image-build.sh -b ubuntu:19.04
```
Comment (arnout) [inline .travis.yml]: Why 19.04?

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 18]
```diff
@@ -8,7 +8,11 @@ dbg() {
 }
 
 err() {
-	echo -e '\033[1;31m'"$@"'\033[0m'
+    printf '\033[1;31m'"$@\n"'\033[0m'
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in:11]: I would have put the newline after the end-of-color, but it doesn't matter much. :ok:

File: common/beerocks/scripts/prplmesh_utils.sh.in [hunk 19]
```diff
@@ -130,17 +134,49 @@ stop_function() {
...
     ps -aux | grep local_bus | grep -v grep
+
+    # check for operational status
+    LOGS_PATH=@BEEROCKS_TMP_PATH@/logs/
+
+    report "Main agent operational" main_agent_operational $LOGS_PATH
+    [ "$error" = "true" ] && cat $LOGS_PATH/beerocks_agent.log
+
+    report "wlan0 radio agent operational" radio_agent_operational $LOGS_PATH wlan0
+    [ "$error" = "true" ] && cat $LOGS_PATH/beerocks_agent_wlan0.log
```
Comment (arnout) [inline common/beerocks/scripts/prplmesh_utils.sh.in]: This will dump the wlan0 log if the first check failed... Which maybe you want, but it is a bit weird. :ok:

File: tools/docker/image-build.sh [hunk 20]
```diff
@@ -21,27 +21,31 @@ scriptdir="$(cd "${0%/*}"; pwd)"
...
-    echo "usage: $(basename $0) [-hvn] <type>"
+    echo "usage: $(basename $0) [-hvbBR] <type>"
     echo "  mandatory:"
     echo "      type - image type <runner/builder>"
     echo "  options:"
     echo "      -h|--help - show this help menu"
     echo "      -v|--verbose - verbosity on"
-    echo "      -b|--base-image - Base OS image to use (Dockerfile 'FROM') "
+    echo "      -b|--base-image - Base OS image to use (Dockerfile 'FROM')"
+    echo "      -B|--builder-name - name for the builder image (default prplmesh-builder)"
```
Comment (arnout) [inline tools/docker/image-build.sh]: What is the use case for setting these names?

I would tend to prefer to use a tag instead, and always use the same tag for both. So `-t <tag>` as the option for the script, and `TAG=":$2"` with TAG defaulting to empty, and `--tag=prplmesh-builder$TAG` in the build call.

File: tools/docker/run.sh [hunk 21]
```diff
@@ -52,18 +53,19 @@ main() {
...
+            -N | --network)     NETWORK="$2"; shift; shift ;;
             -- ) shift; break ;;
             * ) err "unsupported argument $1"; usage; exit 1 ;;
         esac
     done
 
-    docker image inspect prplmesh-runner >/dev/null 2>&1 || {
-        dbg "Image prplmesh-runner does not exist, creating..."
+    docker image inspect ${IMAGE} >/dev/null 2>&1 || {
+        dbg "Image ${IMAGE} does not exist, creating..."
```
Comment (arnout) [inline tools/docker/run.sh]: You'll also have to pass the name to the image-build script, of course.

File: tools/docker/run.sh [hunk 22]
```diff
@@ -52,18 +53,19 @@ main() {
...
-            -h | --help)    usage; exit 0; shift ;;
-            -d | --detach)  DETACH=true; shift ;;
-            -i | --ipaddr)  IPADDR="$2"; shift; shift ;;
-            -n | --name)    NAME="$2"; shift; shift ;;
-            -N | --network) NETWORK="$2"; shift; shift ;;
+            -h | --help)        usage; exit 0; shift ;;
+            -d | --detach)      DETACH=true; shift ;;
+            -i | --ipaddr)      IPADDR="$2"; shift; shift ;;
+            -n | --name)        NAME="$2"; shift; shift ;;
+            -I | --image)  IMAGE="$2"; shift; shift ;;
```
Comment (arnout) [inline tools/docker/run.sh]: This should be tag as well if we decide to go that way, obviously.

**Review by arnout — APPROVED** [review]
Non-essential changes only, or things that can be done later.

File: .travis.yml (hunk 17 above)
Comment (tomereli) [reply .travis.yml]: it stil doesn't work, the agent crashes, i'm trying anything I can think of - it certainly not related to our build.. any idea?

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 19 above)
Comment (tomereli) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: no it isn't, thanks for that

File: common/beerocks/scripts/prplmesh_utils.sh.in (hunk 10 above)
Comment (tomereli) [reply common/beerocks/scripts/prplmesh_utils.sh.in]: This is intentional, we do not map the logs since the host is running multiple instances.

File: .travis.yml (hunk 17 above)
Comment (tomereli) [reply .travis.yml]: wasn't related, it appears that our RELEASE is broken on docker container (worked fine locally for me) - see #93
